// @earendil-works/pi-coding-agent/dist/modes/interactive/components/trust-selector.js
import { Container, getKeybindings, Spacer, Text } from "@earendil-works/pi-tui";
import { theme } from "../theme/theme.js";
import { DynamicBorder } from "./dynamic-border.js";
import { keyHint, rawKeyHint } from "./keybinding-hints.js";
const TRUST_OPTIONS = [
    { label: "Trust", trusted: true },
    { label: "Do not trust", trusted: false },
];
function formatDecision(decision) {
    if (decision === true) {
        return "trusted";
    }
    if (decision === false) {
        return "untrusted";
    }
    return "none";
}
export class TrustSelectorComponent extends Container {
    selectedIndex;
    listContainer;
    savedDecision;
    onSelectCallback;
    onCancelCallback;
    constructor(options) {
        super();
        this.savedDecision = options.savedDecision;
        this.selectedIndex = Math.max(0, TRUST_OPTIONS.findIndex((option) => option.trusted === options.savedDecision));
        this.onSelectCallback = options.onSelect;
        this.onCancelCallback = options.onCancel;
        this.addChild(new DynamicBorder());
        this.addChild(new Spacer(1));
        this.addChild(new Text(theme.fg("accent", theme.bold("Project trust")), 1, 0));
        this.addChild(new Text(theme.fg("muted", options.cwd), 1, 0));
        this.addChild(new Spacer(1));
        this.addChild(new Text(theme.fg("muted", `Saved decision: ${formatDecision(options.savedDecision)}`), 1, 0));
        this.addChild(new Text(theme.fg("muted", `Current session: ${options.projectTrusted ? "trusted" : "untrusted"}`), 1, 0));
        this.addChild(new Spacer(1));
        this.listContainer = new Container();
        this.addChild(this.listContainer);
        this.addChild(new Spacer(1));
        this.addChild(new Text(rawKeyHint("↑↓", "navigate") +
            "  " +
            keyHint("tui.select.confirm", "save") +
            "  " +
            keyHint("tui.select.cancel", "cancel"), 1, 0));
        this.addChild(new Spacer(1));
        this.addChild(new DynamicBorder());
        this.updateList();
    }
    updateList() {
        this.listContainer.clear();
        for (let i = 0; i < TRUST_OPTIONS.length; i++) {
            const option = TRUST_OPTIONS[i];
            if (!option) {
                continue;
            }
            const isSelected = i === this.selectedIndex;
            const isCurrent = option.trusted === this.savedDecision;
            const checkmark = isCurrent ? theme.fg("success", " ✓") : "";
            const prefix = isSelected ? theme.fg("accent", "→ ") : "  ";
            const label = isSelected ? theme.fg("accent", option.label) : theme.fg("text", option.label);
            this.listContainer.addChild(new Text(`${prefix}${label}${checkmark}`, 1, 0));
        }
    }
    handleInput(keyData) {
        const kb = getKeybindings();
        if (kb.matches(keyData, "tui.select.up") || keyData === "k") {
            this.selectedIndex = Math.max(0, this.selectedIndex - 1);
            this.updateList();
        }
        else if (kb.matches(keyData, "tui.select.down") || keyData === "j") {
            this.selectedIndex = Math.min(TRUST_OPTIONS.length - 1, this.selectedIndex + 1);
            this.updateList();
        }
        else if (kb.matches(keyData, "tui.select.confirm") || keyData === "\n") {
            const selected = TRUST_OPTIONS[this.selectedIndex];
            if (selected) {
                this.onSelectCallback(selected.trusted);
            }
        }
        else if (kb.matches(keyData, "tui.select.cancel")) {
            this.onCancelCallback();
        }
    }
}
